### PR TITLE
Fix #1170: Allow customizable width for GitHub Stats Card

### LINKED ISSUE #1170: Add card_width for stats Card repo cards
**Is your feature request related to a problem? Please describe.**
The GitHub Stats Card is too wide and I cannot fit two panels in my `README.md`.

**Describe the solution you'd like**
Allow customizing the width of GitHub Stats Card like it is allowed for the Language Card using card_width.

**Describe alternatives you've considered**
Make the GitHub Stats Card smaller (width) might be also a solution.

**Additional context**
**Current look:**
</img>

**Requested feature:** I would like to achieve something like this (edited in MS Paint):
</img>

**Note:**
This is my 2nd attempt (the 1st one is [here](https://github.com/anuraghazra/github-readme-stats/issues/780)) to draw attention and get feedback as one of my feature requests remained inactive despite the overall activity in the issues section.

### PR DESCRIPTION
**Pull request by postatum**
Fixes #1170 by allowing customizable width for Github Stats Card.

To avoid things overlaping each other, Stats Card has minimum widths which are applied when provided `custom_width` is less than certain values.

### CODE REVIEW

**Comment by rickstaa** [comment]
@postatum Thanks for your pull request. I quickly reviewed your code and it looks good to me. It however looks that the current code doesn't take into account the `show_icons` argument (see last example below). You might have to reload several times to see all the stats since Github will likely timeout trying to request so many stats at the same time.

### hide_rank=true

#### Test without card_width

Rick Staa's GitHub stats

#### Test with to small card_width (i.e 50)

This card width is `220` px smaller than the minimum card width of `270` px.

Rick Staa's GitHub stats

#### Test with card_width 300 

This card width is `30` px bigger than the minimum card width of `270`.

Rick Staa's GitHub stats

####  Test with card_width 600

Rick Staa's GitHub stats

### hide_rank=true, show_icons=true

#### Test without card_width

Rick Staa's GitHub stats

#### Test with to small card_width (i.e 50)

This card width is `220` px smaller than the minimum card width of `270` px.

Rick Staa's GitHub stats

#### Test with card_width 300 

This card width is `30` px bigger than the minimum card width of `270`.

Rick Staa's GitHub stats

####  Test with card_width 600

Rick Staa's GitHub stats

### hide_rank=false

#### Test without card_width

Rick Staa's GitHub stats

#### Test with to small card_width (i.e 50)

This card width is `220` px smaller than the minimum card width of `270` px.

Rick Staa's GitHub stats

#### Test with card_width 300 

This card width is `30` px bigger than the minimum card width of `270`.

Rick Staa's GitHub stats

####  Test with card_width 600

Rick Staa's GitHub stats

### hide_rank=false, show_icons=true

#### Test without card_width

Rick Staa's GitHub stats

#### Test with to small card_width (i.e 50)

This card width is `220` px smaller than the minimum card width of `270` px.

Rick Staa's GitHub stats

#### Test with card_width 300 

This card width is `30` px bigger than the minimum card width of `270`.

Rick Staa's GitHub stats

####  Test with card_width 600

Rick Staa's GitHub stats

File: src/cards/stats-card.js [hunk 1]
```diff
@@ -203,13 +185,22 @@ const renderStatsCard = (stats = {}, options = { hide: [] }) => {
...
 
-  const width = hide_rank
+  /*
+    When hide_rank=true, minimum card width is max of 270 and length of title + paddings.
+    When hide_rank=false, minimum card_width is 340.
+    Numbers are picked by looking at existing dimensions on production.
+  */
+  const minCardWidth = hide_rank
     ? clampValue(
         50 /* padding */ + calculateTextWidth() * 2,
```
Comment (rickstaa) [inline src/cards/stats-card.js:195]: I think here you should also take into account the icon width when the `hide_icons` flag is set to `false`.

File: src/cards/stats-card.js [hunk 2]
```diff
@@ -203,13 +185,22 @@ const renderStatsCard = (stats = {}, options = { hide: [] }) => {
     return measureText(custom_title ? custom_title : i18n.t("statcard.title"));
   };
 
-  const width = hide_rank
+  /*
+    When hide_rank=true, minimum card width is max of 270 and length of title + paddings.
```
Comment (rickstaa) [inline src/cards/stats-card.js:189]: Maybe change this to:

```text 
When hide_rank=true, the minimum card width is 270 px plus the length of title and paddings.
```

The `max` confused me a bit.

**Comment by rickstaa** [comment]
@postatum, please see https://github.com/postatum/github-readme-stats/pull/1/files for a fix to the issue described above. It also trims trialling whitespace, but I can remove that if you want.

File: src/cards/stats-card.js (hunk 2 above)
Comment (rickstaa) [inline src/cards/stats-card.js:189]: Maybe change this to:

```text
When hide_rank=true, the minimum card width is 270 px + the title length and padding.
When hide_rank=false, the minimum card_width is 340 px + the icon width (if show_icons=true).
Numbers are picked by looking at existing dimensions on production.
```

The ``max`` in your old description confused me a bit.

File: src/cards/stats-card.js [hunk 3]
```diff
@@ -203,13 +185,22 @@ const renderStatsCard = (stats = {}, options = { hide: [] }) => {
...
+  const minCardWidth = hide_rank
     ? clampValue(
         50 /* padding */ + calculateTextWidth() * 2,
-        270 /* min */,
-        Infinity,
-      )
-    : 495;
+        270,
+        Infinity)
+    : 340
```
Comment (rickstaa) [inline src/cards/stats-card.js:198]: I think here you should also take into account the icon width when the show_icons flag is set to true. This is 16 px. Please merge https://github.com/postatum/github-readme-stats/pull/1 to fix this issue.

**Review by rickstaa — CHANGES_REQUESTED** [review]

**Comment by anuraghazra** [comment]
The rank circle looks like shifted way too much to the right on the new updates 

Left: THIS PR
Right: CURRENT
image

**Comment by rickstaa** [comment]
@anuraghazra You are right I missed that. This seems to be caused by the following code that was added by @postatum:

https://github.com/anuraghazra/github-readme-stats/blob/dc78bad925e57ac875e9da435c546c1c4a63ab97/src/cards/stats-card.js#L230

I added a fix for this to my pull request.

#### Test with to small card_width (i.e 50)

This card width is `220` px smaller than the minimum card width of `270` px.

Rick Staa's GitHub stats

#### Test with card_width 300 

This card width is `30` px bigger than the minimum card width of `270`.

Rick Staa's GitHub stats

####  Test with card_width 600

Rick Staa's GitHub stats

**Comment by rickstaa** [comment]
Left: THIS PR https://github-readme-stats-git-testwidthfix2-rickstaa.vercel.app/api?username=anuraghazra&show_icons=true&hide_rank=false
Right: CURRENT https://github-readme-stats.vercel.app/api?username=anuraghazra&show_icons=true&hide_rank=false

| <a href="https://github.com/anuraghazra/github-readme-stats">Anurag's github stats</a> | <a href="https://github.com/anuraghazra/github-readme-stats"></a> |
| ------------- | ------------- |

**Comment by anuraghazra** [comment]
@postatum can you merge in the changes done by @rickstaa? 

Also @rickstaa it's better to keep the PR waterfall minimal, try to code review the PR first instead of committing to the branch of Artem.

**Comment by rickstaa** [comment]
@anuraghazra No problem, I will do that next time. I created the pull request to speed up the development process since it solves one of my problems.

**Comment by postatum** [comment]
Hi guys!

Thanks for reviewing this PR. Sorry, I didn't have time to merge your PR. I did it just now.

**Comment by rickstaa** [comment]
@postatum No problem. Thanks again for creating this pull request!

File: tests/renderStatsCard.test.js [hunk 4]
```diff
@@ -75,6 +75,22 @@ describe("Test renderStatsCard", () => {
...
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 400 });
+    expect(document.querySelector("svg")).toHaveAttribute("width", "400");
+  });
+
+  it("should render with custom width set and limit minimum width", () => {
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 1 });
+    expect(document.querySelector("svg")).toHaveAttribute("width", "340");
+
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 1, hide_rank: true });
+    expect(document.querySelector("svg")).toHaveAttribute("width", "305.81250000000006");
```
Comment (rickstaa) [inline tests/renderStatsCard.test.js:91]: @postatum It looks like the current card width with this setting should be 270 px.

Rick Staa's Github stats

File: tests/renderStatsCard.test.js [hunk 5]
```diff
@@ -75,6 +75,22 @@ describe("Test renderStatsCard", () => {
...
+
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 400 });
+    expect(document.querySelector("svg")).toHaveAttribute("width", "400");
+  });
+
+  it("should render with custom width set and limit minimum width", () => {
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 1 });
+    expect(document.querySelector("svg")).toHaveAttribute("width", "340");
+
+    document.body.innerHTML = renderStatsCard(stats, { card_width: 1, hide_rank: true });
```
Comment (rickstaa) [inline tests/renderStatsCard.test.js:90]: @postatum Maybe we should also add tests for the `show_icons` option. My commit now adds `16px` when this option is enabled.

Rick's Github Stats

**Review by rickstaa — CHANGES_REQUESTED** [review]

File: src/cards/stats-card.js [hunk 6]
```diff
@@ -203,13 +184,20 @@ const renderStatsCard = (stats = {}, options = { hide: [] }) => {
...
+  /*
+    When hide_rank=true, the minimum card width is 270 px + the title length and padding.
+    When hide_rank=false, the minimum card_width is 340 px + the icon width (if show_icons=true).
+    Numbers are picked by looking at existing dimensions on production.
+  */
+  const iconWidth = show_icons ? 16 : 0;
+  const minCardWidth = hide_rank
+    ? clampValue(50 /* padding */ + calculateTextWidth() * 2, 270, Infinity)
+    : 340 + iconWidth;
+  const defaultCardWidth = hide_rank ? 270 : 495;
```
Comment (rickstaa) [inline src/cards/stats-card.js:196]: @postatum, @anuraghazra Maybe we should also add the `iconWidth` to the `minwidth` of the card when `hide_rank=true`. By doing this the behaviour for the `hide_icons` option is similar to the case where `hide_rank=false`.

```js
const defaultCardWidth = hide_rank ? 270 + show_icons : 495;
```

**Review by rickstaa — CHANGES_REQUESTED** [review]

**Comment by douglasg14b** [comment]
Is this blocked by code review mostly?

**Comment by rickstaa** [comment]
@douglasg14b Thanks for your interest in this feature. I already reviewed this PR some time ago. I, however, have to check it one last time. There are currently three things preventing this PR from being merged:

    1. https://github.com/anuraghazra/github-readme-stats/pull/1334#discussion_r745146226: A test for the `show_icons` needs to be added.
    2. I did to receive a reaction yet from @anuraghazra and @postatum about https://github.com/anuraghazra/github-readme-stats/pull/1334#discussion_r745148745. I am happy to implement this change, but some feedback would be excellent.
    3. I need to rebase this on top of the main branch. This needs to be carefully done since #1628 changed the code also altered in this PR.

After these points are solved, this PR can be merged. However, I am very short on time since I'm in the final stage of my master's thesis. Therefore, I cannot make a promise when I have time to look at this PR again.

**Comment by rickstaa** [comment]
Sorry for the delay but I just merged this into the master. @postatum Thanks a lot for implementing this feature 🚀.

**Pull request merged**